### ISSUE TITLE
TypeError: Cannot read properties of undefined (reading 'slice')  at fromZodError

### ISSUE THREAD

**Issue by abdulghanitech** [causaly/zod-validation-error#42]
I'm facing this issue. How do we solve this?

```
TypeError: Cannot read properties of undefined (reading 'slice')
    at fromZodError (/Users/abdulghani/Documents/projects/office/wayship-ahoy-client/node_modules/zod-validation-error/dist/cjs/ValidationError.js:71:10)
    ```

**Comment by thanoskrg** [comment]
Hi @abdulghanitech , can you please provide some extra context on the issue above?

Your schema definition, as well as your code snippet that uses `zod-validation-error`, you be very helpful.

**Comment by jmike** [comment]
Closing this issue due to inactivity

**Comment by unRealGamer28** [comment]
Hello, I was testing this library today and I received the same error.

Error:
```
[redacted]\node_modules\zod-validation-error\dist\cjs\ValidationError.js:111
        .slice(0, maxIssuesInMessage)
         ^

TypeError: Cannot read properties of undefined (reading 'slice')
```

Here is my test script:
```
import { z } from "zod";
import { fromZodError, isValidationErrorLike } from "zod-validation-error";

const schema = z.object({
  param1: z.string().toUpperCase(),
  param2: z.union([z.enum(["latest", "all"]), z.number()]),
});

try {
  // throw new Error("test"); // Uncommenting this line throws 'slice' error

  console.table(
    schema.parse({
      param1: "User",
      param2: "alll",
    })
  );
} catch (e) {
  const error = fromZodError(e); // This line throws the error

  if (isValidationErrorLike(error)) {
    console.log("400 : Bad Data (this is a client error)");
    console.log(error.message);
  } else {
    console.log("500 : Server Error");
    console.log(e.message);
  }
}
```

Some things I noted:
1. The above script behaves as expected until 'Error' is thrown (instead of 'ZodError').
2. I need this line `const error = fromZodError(e);` as 'isValidationErrorLike' function takes ValidationError parameter (and not ZodError directly).
3. Making the change below seems to fix this, but there's better ways to fix this depending on the direction of this library:
```
function fromZodError(zodError, options = {}) {
    const { maxIssuesInMessage = MAX_ISSUES_IN_MESSAGE, issueSeparator = ISSUE_SEPARATOR, unionSeparator = UNION_SEPARATOR, prefixSeparator = PREFIX_SEPARATOR, prefix = PREFIX, includePath = true, } = options;

    // if (zodError.errors === undefined) { // Adding this fixes the 'slice' error while maintaining the same functionality in 'isValidationErrorLike' (which would then return false)
    //     return null;
    // }
    
    const reason = zodError.errors
        .slice(0, maxIssuesInMessage)
        .map((issue) => getMessageFromZodIssue({
        issue,
        issueSeparator,
        unionSeparator,
        includePath,
    }))
        .join(issueSeparator);
    const message = conditionallyPrefixMessage(reason, prefix, prefixSeparator);
    return new ValidationError(message, zodError.errors);
}
```

**Comment by thanoskrg** [comment]
I am reopening this issue for further investigation. Thanks @unRealGamer28 for providing more info on this. Can you also please let us know the library version you are using?

**Comment by unRealGamer28** [comment]
Of course! As per my package.json:
"zod-validation-error": "^2.1.0"

**Comment by thanoskrg** [comment]
Thank you @unRealGamer28 

So, taking a better look at this, and going through the [docs](https://github.com/causaly/zod-validation-error?tab=readme-ov-file#fromzoderror), I think that this is the intended way to work for `fromZodError`. It's purpose is to accept `ZodError` instances and convert them to `ValidationError` ones. That's why we always expect a `.errors` property which is of type array and we invoke `.slice` on it. Makes sense?

**Comment by unRealGamer28** [comment]
Hey @thanoskrg, yes that made sense but it doesn't solve the issue I'm facing with my test script. How are we supposed to catch an error and know if it's a ZodError or a normal error before sending it to zod-validation-error?
1. isValidationErrorLike() only works with ValidationError. It doesn't work with ZodError. So I need to use 'fromZodError' to convert ZodError | ValidationError | Error to ValidatorError. Otherwise isValidationErrorLike() throws another error.
2. I don't want a try catch block inside a try catch block.
3. Am I expected to validate if it's a ZodError vs Error myself? I can but it seems like an unnecessary step on my side.

For reference, here is my test script again. Please tell me a way to catch an error and parse it. It may be a ZodError or it may be an ValidationError (as I can manually throw that) or it may be an traditional Error (as some unexpected code may throw that):
```
import { z } from "zod";
import { fromZodError, isValidationErrorLike } from "zod-validation-error";

const schema = z.object({
  param1: z.string().toUpperCase(),
  param2: z.union([z.enum(["latest", "all"]), z.number()]),
});

try {
  // throw new Error("test"); // Uncommenting this line throws 'slice' error

  console.table(
    schema.parse({
      param1: "User",
      param2: "alll",
    })
  );
} catch (e) {
  const error = fromZodError(e); // This line throws the error

  if (isValidationErrorLike(error)) {
    console.log("400 : Bad Data (this is a client error)");
    console.log(error.message);
  } else {
    console.log("500 : Server Error");
    console.log(e.message);
  }
}
```

**Comment by unRealGamer28** [comment]
I saw the referenced issue # 207. And I believe that is a duplicate of this one. This is an issue many people may face when first using zod-validation-error library. They need to validate if it's a ZodError themselves like:
`const error = e instanceof ZodError ? fromZodError(e) : e;`

This is not a problem but an extra step on our side. It may be good to include something like below as an example to README:
```
import { z, ZodError } from "zod";
import { fromZodError, isValidationErrorLike } from "zod-validation-error";

const schema = z.object({
  param1: z.string().toUpperCase(),
  param2: z.union([z.enum(["latest", "all"]), z.number()]),
});

try {
  //   throw new Error("test");

  console.table(
    schema.parse({
      param1: "User",
      param2: "alll",
    })
  );
} catch (e) {
  const error = e instanceof ZodError ? fromZodError(e) : e;

  if (isValidationErrorLike(error)) {
    console.log("400 : Bad Data (this is a client error)");
    console.log(error.message);
  } else {
    console.log("500 : Server Error");
    console.log(e.message);
  }
}
```

You can close this issue if this is what we must do! I'm going to use the code above. Thanks for your help @thanoskrg

**Comment by thanoskrg** [comment]
> They need to validate if it's a ZodError themselves

@unRealGamer28 I understand your concern, although your test script above does only one thing, which is an object parsing using a zod schema. But what if in a real-world scenario your function did many things? Like http requests, IO operations plus zod schema parsing. Then you might end up with all sorts of different error instances. 

So the point I'm trying to make is that it is almost inevitable to do a check against the instance of the error you are getting in the catch block. But there might be another discussion here, whether we could entertain the idea of adding an extra helper in the library, something like `isZodError`. 

cc @jmike @nikoskalogridis what are your thoughts on this?

**Comment by AllusiveBox** [comment]
> > They need to validate if it's a ZodError themselves
> 
> @unRealGamer28 I understand your concern, although your test script above does only one thing, which is an object parsing using a zod schema. But what if in a real-world scenario your function did many things? Like http requests, IO operations plus zod schema parsing. Then you might end up with all sorts of different error instances. 
> 
> So the point I'm trying to make is that it is almost inevitable to do a check against the instance of the error you are getting in the catch block. But there might be another discussion here, whether we could entertain the idea of adding an extra helper in the library, something like `isZodError`. 

I was looking into this library earlier today and came across this problem myself, was about to open an issue before seeing this. I think that it is fine if you guys don't want to change the nature of `fromZodError` to return `null` when given non-Zod errors, but I think throwing a TypeError for trying to call `slice` on `undefined` is the wrong way to go about it. If anything, I would say it should be a `TypeError` for passing an invalid value, because otherwise this feels more like a bug than a feature, especially given that it isn't noted anywhere in the documentation that passing anything other than a `ZodError` breaks the function.

**Comment by thanoskrg** [comment]
Hi @AllusiveBox , thanks for the feedback.

> given that it isn't noted anywhere in the documentation that passing anything other than a `ZodError` breaks the function.

Regarding the above, according to the [documentation](https://github.com/causaly/zod-validation-error?tab=readme-ov-file#arguments-4) `fromZodError` requires an error of `ZodError` instance and is required. 

To be honest, I am under the impression that it's up to the consumer of this function to ensure that they pass proper values, rather than having the function be defensive across any type of input that it might receive. 

But, I am not strong on that, and I would really like to see what's the best way moving forward, and if we need to update the logic, I'm happy to do it. 

I would like to see what are @jmike and @nikoskalogridis thinking of this.

**Comment by AllusiveBox** [comment]
@thanoskrg was just checking to see if there had been any additional discussion regarding this issue that had happened outside this issue or if the intent was still to keep this functionality/documentation the same?

**Comment by thanoskrg** [comment]
Given the number of people that seem to be facing an issue with `.slice` exception when a non-`ZodError` instance is passed to `fromZodError`, we are considering a few options to go about this. 

Before we proceed though, I'd like to ask how come you don't use the more _relaxed_ helper which is `toValidationError`, and internally checks whether it should call `fromZodError` or construct a new `ValidationError` itself. 

For example:

```js
import { toValidationError } from 'zod-validation-error';

try {
  throw new Error('This is an error message');
} catch (err) {
  const validationError = toValidationError({ /* options go here */ })(err);

  console.log(validationError.message); // logs 'This is an error message'
}
```

This is what `toValidationError` does internally:

https://github.com/causaly/zod-validation-error/blob/16055557faf791d18304519105ff03d60e08798c/lib/toValidationError.ts#L8-L18

**Comment by AllusiveBox** [comment]
> Given the number of people that seem to be facing an issue with `.slice` exception when a non-`ZodError` instance is passed to `fromZodError`, we are considering a few options to go about this.
> 
> Before we proceed though, I'd like to ask how come you don't use the more _relaxed_ helper which is `toValidationError`, and internally checks whether it should call `fromZodError` or construct a new `ValidationError` itself.

Honestly? I didn't even see `toValidationError`, likely given the fact that it's not on the npm page, or noted anywhere in the documentation. This does 100% solve the issue I have, however, the lack of any real documentation indicating that this method should be used over the `fromZodError`, or that it even exists. I would almost suggest the removal of `fromZodError` completely and instead recommend that everything go through `toValidationError`, but you guys have `fromZodIssue`, which is helpful as well, and not used in `toValidationError` at all.

If it were me, I'd remove all the `from*` functions (I realize this is a breaking change and therefore wouldn't just happen in the next release) and instead move to a singular `toValidationError` which keeps the current logic, while also adding in logic to convert `ZodIssue`s.

Sidebar, any particular reason for `toValidationError` being a standalone function, rather than a static based method in the `ValidationError` class? Perhaps this is my OOP bias showing, but it seems like it belongs there...

**Comment by jmike** [comment]
>  I didn't even see toValidationError, likely given the fact that it's not on the npm page, or noted anywhere in the documentation

The `toValidationError` has been added to the docs for a while now.
https://github.com/causaly/zod-validation-error?tab=readme-ov-file#tovalidationerror

Point taken that we need to do a better job describing the method.

> Sidebar, any particular reason for toValidationError being a standalone function, rather than a static based method in the ValidationError class?

I am strongly opinionated here against OOP. I personally see no benefit of having it as a static method as part of `ValidationError`. Thanks for the suggestion though.

**Comment by thanoskrg** [comment]
Hey folks, some good news on this one...

We've just pushed a new release, https://github.com/causaly/zod-validation-error/releases/tag/v3.2.0 to tackle the above issues discussed.

TL;DR: 
- You are now able to use `fromError` (instead of `fromZodError`), which is a less strict version of it and we also updated the docs
- We made sure to add a runtime check in `fromZodError` so that it handles a bad use of it and provides a meaningful error to the caller

Update your `package.json` and enjoy! 🤘
